### PR TITLE
Migrate S2053: Implement ShouldExecute (C#)

### LINKED ISSUE #7558: Migrate S2053 to the new Symbolic Execution engine
Migrate `HashesShouldHaveUnpredictableSalt` to the new SE engine.

Split the work up to the following PRs:
* [x] Scaffold the rule
  * Prepare empty rule with `ShouldExecute` false (do the scaffolding for both C# and VB.NET)
  * Prepare tests and test cases - mark old `Noncompliant` as `FIXME Non-compliant` (for C# only)
  * Do not register in `SymbolicExecutionRunner` (yet)
* [x] Implementation (C# only)
  * Implement rule logic, switch `ShouldExecute` to true
  * Register in `SymbolicExecutionRunner`
  * Update ITs
* [x] Finish the C# rule
  * Implement `ShouldExecute`  (C# only)
  * Remove all `FIXME` from test cases, and mark them as FNs or Compliant

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Fixes #7558

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs [hunk 1]
```diff
@@ -26,5 +26,27 @@ public sealed class HashesShouldHaveUnpredictableSalt : HashesShouldHaveUnpredic
...
+
+    private sealed class VulnerableTypeWalker : SafeCSharpSyntaxWalker
+    {
+        public bool Result { get; private set; }
+
+        public override void Visit(SyntaxNode node)
+        {
+            if (!Result && node is IdentifierNameSyntax identifier)
+            {
+                Result = VulnerableTypes.Any(x => identifier.NameIs(x.TypeName));
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs:44]: Will this work if there's a class `ChildPasswordDeriveBytes : PasswordDeriveBytes`?
I saw that it's not sealed.

Is there a UT for this?

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs [hunk 2]
```diff
@@ -26,5 +26,27 @@ public sealed class HashesShouldHaveUnpredictableSalt : HashesShouldHaveUnpredic
...
+        return walker.Result;
+    }
+
+    private sealed class VulnerableTypeWalker : SafeCSharpSyntaxWalker
+    {
+        public bool Result { get; private set; }
+
+        public override void Visit(SyntaxNode node)
+        {
+            if (!Result && node is IdentifierNameSyntax identifier)
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs:42]: Isn't it more specific to check for ObjectCreation syntax?

**Review by mary-georgiou-sonarsource — CHANGES_REQUESTED** [review]
Questions mostly

**Comment by zsolt-kolbay-sonarsource** [comment]
The RSPEC has been updated by AppSec, but the title wasn't changed by mistake.
The specs need to be clarified before this can be merged. Until then, the PR remains a Draft.

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs [hunk 3]
```diff
@@ -26,5 +26,27 @@ public sealed class HashesShouldHaveUnpredictableSalt : HashesShouldHaveUnpredic
...
+        {
+            if (node is ObjectCreationExpressionSyntax || ImplicitObjectCreationExpressionSyntaxWrapper.IsInstance(node))
+            {
+                Result = true;
+            }
+            else
+            {
+                base.Visit(node);
+            }
+        }
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs:50]: There's an explicit visitor for object creation:
```suggestion
        public override void Visit(SyntaxNode node)
        {
            if (!Result)
            {
                base.Visit(node);
            }
        }

        public override void VisitObjectCreationExpression(ObjectCreationExpressionSyntax node) =>
            Result = true;
```

Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs:50]: We're using a lower version of Roslyn, so there's no `VisitXXX` method `ImplicitObjectCreation`.

Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs]: I've moved the `ObjectCreation` to a separate `Visit` method and left `ImplicitObjectCreation` in the default `Visit` method.

**Comment by zsolt-kolbay-sonarsource** [comment]
Got automatically closed by Github after force pushing and wasn't able to re-open it. Even after trying [this solution](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c).

**Pull request closed without merging**